// @chakra-ui/react/dist/cjs/components/file-upload/namespace.cjs
"use strict";
'use strict';

var fileUpload = require('./file-upload.cjs');



exports.ClearTrigger = fileUpload.FileUploadClearTrigger;
exports.Context = fileUpload.FileUploadContext;
exports.Dropzone = fileUpload.FileUploadDropzone;
exports.DropzoneContent = fileUpload.FileUploadDropzoneContent;
exports.FileText = fileUpload.FileUploadFileText;
exports.HiddenInput = fileUpload.FileUploadHiddenInput;
exports.Item = fileUpload.FileUploadItem;
exports.ItemContent = fileUpload.FileUploadItemContent;
exports.ItemDeleteTrigger = fileUpload.FileUploadItemDeleteTrigger;
exports.ItemGroup = fileUpload.FileUploadItemGroup;
exports.ItemName = fileUpload.FileUploadItemName;
exports.ItemPreview = fileUpload.FileUploadItemPreview;
exports.ItemPreviewImage = fileUpload.FileUploadItemPreviewImage;
exports.ItemSizeText = fileUpload.FileUploadItemSizeText;
exports.Items = fileUpload.FileUploadItems;
exports.Label = fileUpload.FileUploadLabel;
exports.List = fileUpload.FileUploadList;
exports.PropsProvider = fileUpload.FileUploadPropsProvider;
exports.Root = fileUpload.FileUploadRoot;
exports.RootProvider = fileUpload.FileUploadRootProvider;
exports.Trigger = fileUpload.FileUploadTrigger;
